### ISSUE TITLE
ArrayIndexOutOfBoundsException in addMarkers(@NonNull List<MarkerOptions> markerOptionsList)

### ISSUE THREAD

**Issue by mrfaa** [mapbox/mapbox-gl-native#4487]
The method 

> public List<Marker> addMarkers(@NonNull List<MarkerOptions> markerOptionsList)

 in the class MapboxMap.java throws an 

> java.lang.ArrayIndexOutOfBoundsException: length=0; index=0

 although size of markerOptionsList is not 0.

My Mapbox Version is com.mapbox.mapboxsdk:mapbox-android-sdk:4.0.0-beta.2@aar

**Comment by bleege** [comment]
@mrfaa Thanks for the feedback!  Is this still happening with the RC 1 build from Friday?

**Comment by jfirebaugh** [comment]
@mrfaa Can you please provide the complete stack trace, and if possible, the code that calls `addMarkers`?

**Comment by tobrun** [comment]
I have been looking into but I'm unable to reproduce locally. I tried reproducing this with an empty list, single item list and multiple items in a list in the test app. I also tried triggering this behaviour through unit tests but fail to reproduce.

The only use-case that I can currently think of that could trigger such a crash is that the `NativeMapView` is being destroyed while we are trying to add markers to it. The `mMapView.addMarkers(markers)` will return an `empty long[]`. Afterwards we are not checking if the size of markers is the same as size of the empt list

```java
  long[] ids = mMapView.addMarkers(markers);
  long id = 0;
  Marker m;
  for (int i = 0; i < markers.size(); i++){
       ...
       id = ids[i];
       ...
```


I currently added this check before iterating the markers:

```java
// if unittests or markers are correctly added to map
if (ids == null || ids.length == markers.size()) {
```

The `ids=null` check is needed for unit tests, the second part validates if the markers were correctly added on NativeMapView.

**Comment by tobrun** [comment]
Now also noticing that I need to do a similar fix for `Polylines` and `Polygons`.

**Comment by mrfaa** [comment]
Thanks for the feedback. I could not reproduce the ArrayIndexOutOfBoundsException with the RC1 Build from Friday.

But for further investigations here is the complete stack trace:
```
Fatal Exception: java.lang.ArrayIndexOutOfBoundsException: length=0; index=0
       at com.mapbox.mapboxsdk.maps.MapboxMap.addMarkers(MapboxMap.java:629)
       at xxx.MapOperations.populateMapWithAllTags(MapOperations.java:61)
       at xxx.activity.MainActivity$6.onLocationChanged(MainActivity.java:408)
       at xxx.util.LocationProvider$1.onLocationChanged(LocationProvider.java:65)
       at android.location.LocationManager$ListenerTransport._handleMessage(LocationManager.java:255)
       at android.location.LocationManager$ListenerTransport.access$000(LocationManager.java:184)
       at android.location.LocationManager$ListenerTransport$1.handleMessage(LocationManager.java:200)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:176)
       at android.app.ActivityThread.main(ActivityThread.java:5419)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:525)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1046)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:862)
       at dalvik.system.NativeStart.main(NativeStart.java)

```

The method which calls addMarkers:
`mapboxMap.addMarkers(markerList);`
with
`private final List<MarkerOptions> markerList = new ArrayList<>();`
the markerList is filled with MarkerOptions marker.

**Comment by bleege** [comment]
Rebased and Merged into `release-ios-3.2.0-android-4.0.0`.

Thanks again for reporting this @mrfaa!

**Comment by mrfaa** [comment]
Thanks for fixng this issue!
